Add Total Ordering algorithms to documentation (#25)
optakt/optakt.github.io: content/flare/consensus.md

Maintaining the consistency of replicated information is an important and problematic aspect of the implementations, and it can be simplified by imposing a total order, rather than a partial order, on messages.
The total order on messages ensures that exactly the same updates are applied to the replicated data in the same order at all nodes.

The Total Ordering algorithms are executed independently and concurrently by each process and each of them incrementally creates a total order, without requiring the transmission of any additional messages in the general case.
They operate in the presence of asynchrony and communication faults, which may cause different processes to receive messages in different orders, and they tolerate both crash and Byzantine process fault.
The system is asynchronous in the sense that no bound can be placed on the time required for a computation or for a communication of a message.
A crash fault causes a process to stop processing and generating messages.

> A Byzantine fault allows an arbitrary behavior by the faulty process, including generating of messages intended to mislead the other processes.

Author: I rephrased it as you suggested https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R308
I think is cleared now as well

Each broadcast message has a header consisting of a process identifier, a message sequence number and payload content.
A typical implementation might use a digital signature to substantiate the source of a message (process identifier) for each node.
Systems capable of tolerating Byzantine faults must be able to confirm the source of a message and to detect messages that have the same header but different contents and/or acknowledgements within the message.
Another option is to also embed a digest of the message content in its identifier, so that acknowledgements of messages with different contents but same sequence number will be regarded as acknowledgements of distinct messages.
The acknowledgement and re-transmission mechanisms can ensure that all versions of a message are delivered to all destinations.
Digital signatures and digests are computationally expensive, but do not require the additional rounds of message exchange of alternative strategies.

The output of the algorithms is a total order of messages that is identical at all non-faulty processes in the system.

> Messages from non-Byzantine processes follow all previously broadcast messages. The relation between messages is transitive, and since each message also follows itself, it is reflexive as well. Messages need to acknowledge previously broadcast messages to be considered non-faulty.

Author: Yes, corrected this here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R320-R322
Thanks!

Messages in the Byzantine casual order prefix (except those in a cycle) provide votes on the candidate sets.
These votes are not contained explicitly in the messages, but are deduced from the partial relationship between messages.
Messages relation in most of the cases is represented by a DAG structure, so votes are implicitly counted using edges between vertices from each position towards its ancestry and progeny.
A non-faulty process decides to advance a candidate set to the total order based on the votes of the messages in its partial order prefix and each process makes its own decisions independently.
Even though the messages can be delayed or lost, every non-faulty process must decide on the same message by which to extend to the total order.

An _execution step_ consists of adding one message to the Byzantine casual (partial) order prefix and executing the ordering algorithms.
In a step, all candidate sets that can be constructed from the candidate messages in the casual order prefix are considered.

> Examples and further specifications of [Total Ordering algorithms](https://core.ac.uk/display/82213948) specify conditions for each one and calculates its communication complexity and performance.

Author: I rephrased it here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R356
WDYT?
Or you meant to rephrase it in some other way?

The asynchronous nature and faulty behavior of the process and the communication medium are reflected in the Byzantine partial order of messages that is input to the algorithms.

### Aleph Algorithm

One of the examples of Total Order algorithms with total ordering execution is [Aleph](https://arxiv.org/abs/1908.05156).
Aleph is a leaderless, fully asynchronous, Byzantine Fault Tolerant consensus protocol for total ordering of messages that are exchanged between entities in a network.
Consequently, it is based on a distributed construction of a partially ordered set and the execution step - the algorithm for reaching a consensus on its extension to a total order.
To achieve a consensus, the processes perform computations based on only a local copy of the data structure, however, they are bound to end up with the same result as required by the total ordering process.

The nodes in Aleph protocol network send arbitrary messages to each other using a combination of multicast and random gossip for communication.

> Each node knows the identity of all other nodes and identifies itself through its public key, for which it holds a private key.

Author: Corrected https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R308
Thanks!

> Each node knows the identity of all other nodes and identifies itself through its public key, for which it holds a private key.

Author: Corrected https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R372
Thanks!

The consensus separates the network communication from the protocol logic.
The only role for the network layer is to maintain a common data structure in the form of a DAG tree among all nodes.
The purpose of using it is to divide the algorithm execution into separate virtual rounds.
In each round the nodes emits exactly one "unit" that should be thought of as a message broadcast to all the other nodes, and it is a vertex position on its local DAG version.
Every unit contains information about its creator, its parents and additional data that can be included in it.
All the nodes are expected to generate such units and maintain their local copies of the common DAG, to which new units are continuously added.
Aleph uses transitive properties of the underlying DAG.
Each unit has a "DAG-round" that is defined by the maximum length of a downward chain starting from the current transaction, meaning getting as long as possible path of its ancestors.
A unit with no parents has a DAG-round of `0`, otherwise it has a DAG-round equal to the maximum of DAG-rounds of its parents plus `1`.

> Evey node should create one unit in every round, and it should do so after learning a large portion (at least `2f+1`) of units created in the previous round.

Author: Corrected https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R399
Thanks!

> Then the protocol is stated entirely through [combinatorial properties of this structure](https://cardinal-cryptography.github.io/AlephBFT/how_alephbgt_does_it.html)

Author: I put the wrong link https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R401
Corrected here, thanks!

There is no requirement for a trusted dealer to distribute certain cryptographic keys among nodes, but Aleph implements an ABFT Randomness Beacon with a trustless setup.

Aleph protocol is an example of a Total Ordering algorithm that creates a total order of blocks, thus creating a finalized chain at run, and keep its DAG history records only locally.
This gives to the nodes a higher level of independence when executing the algorithm, and the weight of synchronizing towards reaching the same total order of blocks is transferred to the combinatoric selection of messages from the Byzantine flow.

### Blockmania

Blockmania is also a leaderless Byzantine Fault Tolerant consensus protocol, detailed in the [Blockmania academic paper](https://arxiv.org/abs/1809.01620).
It is a part of the [Chainspace platform](https://chainspace.io/docs/#introduction), but can be used on its own.
Blockmania is another example of a Total Ordering algorithm.

Typically, clients are external to the network nodes and emit transactions that need to be agreed upon.

Nodes in the network maintain an asymmetric signature scheme and all others can authentically verify their messages via public keys.

> Blockmania keep consistent blocks history using a DAG structure, that is subsequently interpreted by each node separately.

Author: Corrected https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R419
Thanks!

Each block emitted by a node is first a _candidate_ block and contains a position defined by its creator node, its sequence number and its content.
Each entry in the block content is either a transaction received from a client or a reference to the hash of a valid block received from another node.
The full block is signed using the private key of its creator and broadcast to all other nodes.

Besides broadcasting blocks, nodes listen to requests from other nodes for specific blocks, and respond by sending them the full contents of the blocks requested.

An honest node includes all transactions received from clients as entries into a block it emits.
Also, all honest nodes receive and checks for validity all blocks they directly or indirectly reference in their created blocks.
This means that an honest node receives and stores locally a copy of the full DAG record, starting from the genesis vertex until the last block it emits.
And if an honest node emits a block, all honest nodes eventually receive and consider it valid (after validating) as well as all blocks that it references directly or indirectly in its content entries.

> A dishonest node may contradict itself and send two distinct blocks for the same position in the chain or not send a block for a position for some time or ever.

Author: Yes, this is fixed here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R436
Thanks!

All honest nodes reference all valid blocks in their own blocks, including contradictory ones, and proceed with the next phase of the protocol called "interpretation".

> The aim of the interpretation phase is for all honest nodes, despite holding a different subset of the DAG record at any point, to eventually agree on a specific block for any position or alternatively to agree that there is no such block, and to produce the next position in the total ordering of blocks.

Author: Foxed here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R439
Thanks!

For reaching a decision for a single position, Blockmania runs an abstract terminating reliable broadcast protocol that is never materialized in real exchanged messages but rather the structure of the block DAG is interpreted by each node.
At any time, a node can emit a message to propose a block in a _pre-prepare_ message to all nodes.
Honest nodes only propose one block for any given position.
This message contains a _view_ value that represents the change in the blockchain state and has a sequence number starting from zero.
Upon receiving the _pre-prepare_ message, nodes that have the same view broadcast a _prepare_ message that contain their identity, view sequence number, block hash and content.
If the view sequence number is zero, meaning this is the first proposed view change, the proposer signature is checked before accepting message.
If the node does not have the same view interpreted, it does not send out _prepare_ message, but sets the received _pre-prepare_ and _prepare_ messages for that view in its input records buffer.

> Upon receiving `2t` _prepare_ messages for a view and a corresponding _pre-prepare_ message in its input buffer, nodes broadcast _commit_ messages for that view.

Author: `t` in this example is some arbitrary number, that is needed to express the ratio.
Did you mean this not clearly explained, or just asked?

After receiving `2t+1` number of _commit_ messages (also saved inside the input buffer), nodes consider view with the next position "decided" and update their local DAG versions.

The consensus assumes that if two honest nodes reach a decision about a position in the chain, defined by block creator and a block sequence number, they always reach the same decision, assuming at most `f` byzantine nodes.
A decision is eventually reached for any position in the total order of blocks by all honest nodes.

The Blockmania considers a reliable transmission between honest nodes to be a Byzantine network that may arbitrarily delay blocks, but where they eventually get delivered.

> This is achieved in practice using re-transmissions.

Author: Yes, fixed here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R456
Thanks!

Also, an honest node may request missing blocks from other nodes, when those have been included in their chains as valid.
Nodes use the gossip protocol to communicate with each other and do not need to know all participants in the network to reach to a decision.

The goal of Blockmania protocol is to ensure that all honest nodes arrive at the same ordering of blocks, despite the presence of Byzantine nodes.
Blockmania may be used with a Proof of Stake system by dynamically defining the quorum of nodes that reach agreement on a block before committing (`t`), thus two honest parties with a partial view of the Block DAG, can reach an agreement on which blocks will be accepted by all.

In the Blockmania consensus, nodes update their block history using a DAG tree available to all, parts of which they just keep as copies in their local memory.
The consensus is built in a modular manner and combines the quorum-based reaching of agreement with the total ordering creation, meaning it is one step further from PBFT types of algorithms.

### SafetyScore

> [SafetyScore consensus](https://safetyscore.app/whitepaper#kairos) is an example of an algorithm that builds a total ordered ledger, that is inspired by the idea of an epidemic spread of a virus that is hard to detect.

Author: Updated here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R469
Thanks!

It is a decentralized protocol that disseminates risk measures among the network participants based on their interactions with a disease.

> It is basically a digital tracing algorithm that can be used from all kinds of applications but preventing leaks of participants personal data in the network.

Author: Yes, this is much better btw; I understand it better when I read it now
Fixed here https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R469
Thanks!

Nodes that can validate blocks published on the distributed ledger are known as validator nodes.
The right to become a validator is earned by nodes that are operating reliably and are as far away as possible from each other in order to increase the overall resilience of the network.

Nodes use asymmetric model of digital signatures that are used for authentication of messages they transmit and use shared secrets derived from ECDG X25519.

In SafetyScore certain organizations are able to run specialized nodes that act as a distributed notary.
They perform roles similar to notaries in the real world and act as witnesses ensuring the documents and transactions are properly executed.
The right to operate a Distributed Notary is limited to organizations that have had a track record of working to preserve digital rights and freedoms, e.g. Electronic Frontier Foundation, Mozilla Foundation, Open Rights Group, etc.
The distributed notary nodes have access to some confidential information like link between two activity keys and the operators should push back against parties looking to de-anonymize users.

> Participants in the distributed ledger entity also enforce punishments to network nodes that violate the protocol, after a valid claim for the faulty act is presented.

Author: Yes, its still distributed notary entity, fixed https://github.com/optakt/optakt.github.io/pull/25/commits/d3ef28d81b0ae48953d2a1bf37127a549d555f46#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R488
Thanks!

Each broadcast message has a header consisting of a process identifier, a message sequence number and payload content.
A typical implementation might use a digital signature to substantiate the source of a message (process identifier) for each node.
Systems capable of tolerating Byzantine faults must be able to confirm the source of a message and to detect messages that have the same header but different contents and/or acknowledgements within the message.
Another option is to also embed a digest of the message content in its identifier, so that acknowledgements of messages with different contents but same sequence number will be regarded as acknowledgements of distinct messages.
The acknowledgement and re-transmission mechanisms can ensure that all versions of a message are delivered to all destinations.
Digital signatures and digests are computationally expensive, but do not require the additional rounds of message exchange of alternative strategies.

The output of the algorithms is a total order of messages that is identical at all non-faulty processes in the system.

> Message `x1` from process `a` follows a message `x0` from process `b` if `x1` acknowledges `x0` or `x1` acknowledges some other message `y` that follows `x0`, where `y` is not the same as `x0` and `y` is not the same as `x1`.

Reviewer: This sentence kind of lost me 😄 I had to read it twice to get it. Maybe this is the kind of situation where a small diagram would be clearer than a sentence? WDYT?

If a Byzantine process ever participates in a cycle, then none of its subsequent messages will be ordered.

A Byzantine process can originate two or more concurrent messages with the same header, but with different contents and/or acknowledgments, neither of which follows the other and possibly each of which purports to follow different messages.
These messages are called _mutants_.

The input to the Total Ordering algorithms is a Byzantine partial order of messages and the output is a total order of messages.
Only information derived from the Byzantine casual order it used to construct the total order, meaning that no additional communication between processes is required.

A "candidate message" is a message from the process's Byzantine casual order that is not yet in the total order but only follows a message from it.
A set of candidate messages is called a "candidate set".
Thus, messages that occur in a cycle cannot be candidate messages, however mutant messages that are not in a cycle can be candidate messages and even members of the same candidate set.

> Except for mutants, at most one candidate message from a given process will be considered for the next advancement to the total order.

Reviewer: ```suggestion
Except for mutants, at most one candidate message from a given process is considered for the next advancement to the total order.
```

Author: Fixed here https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R356
Thanks!

The set of transactions that each one sees is a growing blockchain that does not have a built-in notion of finality and the streams of data are an advancing sequence of blockchain "tips" that each of the participants sees.

The problem of ordering transactions in asynchronous types of protocols, where they are received with some time delay and in mixed order, is also known as _Atomic Broadcast_.
It describes the situation where a set of nodes commonly constructs a total ordering of a set of transactions, where these transactions might not arrive all at the same time.
Every time a node receives a transactions, it processes it and places it at its position.
The sequence of transactions must be output in an incremental manner, meaning that before putting a transaction at some position, previous positions must have their respective transactions set.
It is assumed that whenever an honest node outputs a transaction at some position, every other honest node will also eventually output the same transaction for this position.
Also, every transaction that is input at some honest node is eventually output by all honest nodes.

> This ensures that not only all honest nodes will produce the same ordering, but also that no transactions is lost due to censorship because all honest nodes will output it.

Reviewer: ```suggestion
This ensures that not only all honest nodes produce the same ordering, but also that no transaction is lost due to censorship because all honest nodes output it.
```

Author: Corrected here https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R400
Thanks!

The protocol deals with an expected latency for each transaction, which is the time it is output by all honest nodes after it has been input to at least some preset number of nodes.

The consensus separates the network communication from the protocol logic.
The only role for the network layer is to maintain a common data structure in the form of a DAG tree among all nodes.
The purpose of using it is to divide the algorithm execution into separate virtual rounds.
In each round the nodes emits exactly one "unit" that should be thought of as a message broadcast to all the other nodes, and it is a vertex position on its local DAG version.
Every unit contains information about its creator, its parents and additional data that can be included in it.
All the nodes are expected to generate such units and maintain their local copies of the common DAG, to which new units are continuously added.
Aleph uses transitive properties of the underlying DAG.
Each unit has a "DAG-round" that is defined by the maximum length of a downward chain starting from the current transaction, meaning getting as long as possible path of its ancestors.

> A unit with no parents will have a DAG-round 0, otherwise it has a DAG-round equal to the maximum of DAG-rounds of its parents plus one.

Reviewer: ```suggestion
A unit with no parents has a DAG-round of `0`, otherwise it has a DAG-round equal to the maximum of DAG-rounds of its parents plus `1`.
```

Author: Fixed here https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R414
Thanks!

Evey node should create one unit in every round, and it should do so after learning a large portion (at least `2f+1`) of units created in the previous round.

Then the protocol is stated entirely through [combinatorial properties of this structure](https://cardinal-cryptography.github.io/AlephBFT/how_alephbgt_does_it.html)
Each of the nodes votes on the units it sees and decisions on accepting it or not are binary.

> In a network running Aleph protocol, where a steady inflow of transactions is being assumed, the communication complexity achieved is O(N² logN).

Reviewer: ```suggestion
In a network running Aleph protocol, where a steady inflow of transactions is assumed, the communication complexity achieved is `O(N² logN)`.
```

Author: Fixed https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R420
Thanks, looks better now!

> There is no requirement for a trusted dealer to distribute certain cryptographic keys among nodes, but Aleph rather implements an ABFT Randomness Beacon with trustless setup.

Reviewer: ```suggestion
There is no requirement for a trusted dealer to distribute certain cryptographic keys among nodes, but Aleph implements an ABFT Randomness Beacon with a trustless setup.
```

Author: Corrected https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R421
Thanks!

Blocks from each node form a sequence and each refers to the hash of the previous one, forming a hash chain.
A block contains information about its creator node, its sequence number in the chain, a sequence of entries representing the content of the block.
Each block emitted by a node is first a _candidate_ block and contains a position defined by its creator node, its sequence number and its content.
Each entry in the block content is either a transaction received from a client or a reference to the hash of a valid block received from another node.
The full block is signed using the private key of its creator and broadcast to all other nodes.

Besides broadcasting blocks, nodes listen to requests from other nodes for specific blocks, and respond by sending them the full contents of the blocks requested.

An honest node includes all transactions received from clients as entries into a block it emits.
Also, all honest nodes receive and checks for validity all blocks they directly or indirectly reference in their created blocks.
This means that an honest node receives and stores locally a copy of the full DAG record, starting from the genesis vertex until the last block it emits.

> And if an honest node emits a block, all honest nodes will eventually receive and consider valid (after validating) this block as well as all blocks that it references directly or indirectly in its content entries.

Reviewer: ```suggestion
And if an honest node emits a block, all honest nodes eventually receive and consider it valid (after validating) as well as all blocks that it references directly or indirectly in its content entries.
```

Author: Fixed https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R450
Missed the future tense in that place, thanks!

A dishonest node may contradict itself and send two distinct blocks for the same position in the chain or not send a block for a position for some time or ever.
All honest nodes reference all valid blocks in their own blocks, including contradictory ones, and proceed with the next phase of the protocol called "interpretation".

The aim of the interpretation phase is for all honest nodes, despite holding a different subset of the DAG record at any point, to eventually agree on a specific block for any position or alternatively to agree that there is no such block, and to produce the next position in the total ordering of blocks.
Blockmania is created as a simplified variant of PBFT algorithm and all parties need to agree on a single position rather than the sequence and all nodes perform their interpretation process independently.

For reaching a decision for a single position, Blockmania runs an abstract terminating reliable broadcast protocol that is never materialized in real exchanged messages but rather the structure of the block DAG is interpreted by each node.

> At any time, a node can emit a message to propose a block in a "pre-prepare" message to all nodes.

Reviewer: Who's in charge of naming at Blockmania?! 😄 `Pre-prepare` is terrible :p

Author: Yes it is, these are three types of messages in their algorithm steps - pre-prepare, prepare, commit.
To be honest, I was trying hard to avoid mentioning them, but in the end decided to include as brief as possible description of these steps because it seemed reasonable to me to explain at least this

Honest nodes only propose one block for any given position.
This message contains a _view_ value that represents the change in the blockchain state and has a sequence number starting from zero.
Upon receiving the "pre-prepare" message, nodes that have the same view broadcast a "prepare" message that contain their identity, view sequence number, block hash and content.
If the view sequence number is zero, meaning this is the first proposed view change, the proposer signature is checked before accepting message.
If the node does not have the same view interpreted, it does not send out "prepare" message, but sets the received "pre-prepare" and "prepare" messages for that view in its input records buffer.
Upon receiving `2t` "prepare" messages for a view and a corresponding "pre-prepare" message in its input buffer, nodes broadcast "commit" messages for that view.

> After receiving `2t+1` number of "commit" messages (also saved inside the input buffer), nodes consider view with the next position "decided" and update their local DAG versions.

Reviewer: Maybe instead of quoting the message names, we could write them in italics? WDYT? That might make this document easier to process as a reader.

Author: Yep

Author: Fixed message names everywhere

> The consensus assumes that if two honest nodes reach a decision about a position in the chain, defined by block creator and a block sequence number, that decision will be the same, assuming at most `f` byzantine nodes.
> A decision will eventually be reached for any position in the total order of blocks by all honest nodes.

Reviewer: ```suggestion
The consensus assumes that if two honest nodes reach a decision about a position in the chain, defined by block creator and a block sequence number, they always reach the same decision, assuming at most `f` byzantine nodes.
A decision is eventually reached for any position in the total order of blocks by all honest nodes.
```

Author: Corrected both
https://github.com/optakt/optakt.github.io/pull/25/commits/e51539edef429ee0583f05806acf8610dd0971a0#diff-af4a1f964639da79798e8b30e7b99b5fb3f29ea0d4da47acd88f0d9c9867be13R468

> The Blockmania considers a reliable transmission between honest nodes — a Byzantine network may delay arbitrarily blocks, but eventually they will be delivered.

Reviewer: ```suggestion
The Blockmania considers a reliable transmission between honest nodes to be a Byzantine network that may arbitrarily delay blocks, but where they eventually get delivered.
```

I assume this is what you meant?

The total order ensures consistency by having the same updates applied to the replicated pieces of data in the same order at all sites.

Each broadcast message has a header consisting of a process identifier, a message sequence number and payload content.
A typical implementation might use a digital signature to substantiate the source of a message (process identifier) for each node.
Systems capable of tolerating Byzantine faults must be able to confirm the source of a message and to detect messages that have the same header but different contents and/or acknowledgements within the message.
Another option is to also embed a digest of the message content in its identifier, so that acknowledgements of messages with different contents but same sequence number will be regarded as acknowledgements of distinct messages.
The acknowledgement and re-transmission mechanisms can ensure that all versions of a message are delivered to all destinations.
Digital signatures and digests are computationally expensive, but do not require the additional rounds of message exchange of alternative strategies.

> The output of the algorithms is a total order of messages that is identical at all non-faulty processes in the system.
> Message `x1` from process `a` follows a message `x0` from process `b` if `x1` acknowledges `x0` or `x1` acknowledges some other message `y` that follows `x0`, where `y` is not the same as `x0` and `y` is not the same as `x1`.
> This implies that the relation between messages is transitive and also each message follows itself, consequently the relation is also reflexive.
> A message `x1` from a non-Byzantine process `a` acknowledges a message `x0` only if the process has received all messages that `x0` follows before it broadcast `x1`.
> 
> ```text
>  a       b                chain
>  |       |                  |
>  |       |                  |
>  |       |                  |
>  |      `x0`------------->  |  ↓
>  |                          |
>  |     [`y`-------------->] |  ↓
>  |                          |
> `x1`--------------------->  |  ↓
>                             |
>                            ...
>  ```
> 
> Consequently, a message from a non-Byzantine process follows all messages broadcast earlier from that process.

Reviewer: Actually after thinking about it more, I think the explanation could be simplified. I don't know how we could represent this as a diagram without actually making it more confusing than it should be 😄 WDYT of:

```suggestion
The output of the algorithms is a total order of messages that is identical at all non-faulty processes in the system.

Messages from non-Byzantine processes follow all previously broadcast messages. The relation between messages is transitive, and since each message also follows itself, it is reflexive as well. Messages need to acknowledge previously broadcast messages to be considered non-faulty.
```

I think this is enough to convey the idea that messages need to acknowledge prior messages in order to be considered valid? WDYT?

Author: Looks good to me, uodated here https://github.com/optakt/optakt.github.io/pull/25/commits/7a07360c1ddc79a9eff45017b245bf1f89997096
Thanks!
